test/jdk/jdk/internal/util/VersionTest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) 2022, 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+
+import java.util.stream.Stream;
+
+import jdk.internal.util.Version;
+
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.fail;
+
+/**
+ * @test
+ * @summary test jdk.internal.util.Version
+ * @modules java.base/jdk.internal.util
+ * @run junit VersionTest
+ */
+
+public class VersionTest {
+
+    private static Stream<Arguments> versionParams() {
+        return Stream.of(
+                Arguments.of("1", new Version(1, 0)),
+                Arguments.of("1.2", new Version(1, 2)),
+                Arguments.of("1.2", new Version(1, 2, 0)),
+                Arguments.of("1.2.3", new Version(1, 2, 3)),
+                Arguments.of("1-abc", new Version(1, 0, 0)), // Ignore extra
+                Arguments.of("1.2-abc", new Version(1, 2, 0)), // Ignore extra
+                Arguments.of("1.2.3.4", new Version(1, 2, 3)), // Ignore extra
+                Arguments.of("1.2.3-abc", new Version(1, 2, 3)) // Ignore extra
+        );
+    }
+
+    @ParameterizedTest
+    @MethodSource("versionParams")
+    public void checkParse(String verName, Version expected) {
+        Version actual = Version.parse(verName);
+        assertEquals(actual, expected, "Parsed version mismatch");
+    }
```

Comment:
Another way to declare arguments is to use `CsvSource` like:

```java
    @ParameterizedTest
    @CsvSource(textBlock =
            """
            1, 0, 0, 1
            1, 2, 0, 1.2
            1, 2, 0, 1.2.0
            1, 2, 3, 1.2.3
            # Ignore extra trailing characters
            1, 0, 0, 1-abc
            1, 2, 0, 1.2-abc
            1, 2, 3, 1.2.3.4
            1, 2, 3, 1.2.3-abc
            """)
    void checkParse(int major, int minor, int micro, Version actual) {
        Version expected = new Version(major, minor, micro);
        assertEquals(expected, actual, "Parsed version mismatch");
    }
```

Reply:
Good to know about CsvSource.
The test is checking Version.parse(str) works. So I'd code it as:
```
void checkParse(int major, int minor, int micro, String verString) {
        Version expected = new Version(major, minor, micro);
        Version actual = Version.parse(verString);
        assertEquals(expected, actual, "Parsed version mismatch");
    }
```